app/src/main/java/org/thoughtcrime/securesms/util/FlowUtils.kt

```diff
@@ -0,0 +1,42 @@
+package org.thoughtcrime.securesms.util
+
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.flatMapConcat
+
+/**
+ * Buffers items from the flow and emits them in batches. The batch will have size [maxItems] and
+ * time [timeoutMillis] limit.
+ */
+fun <T> Flow<T>.timedBuffer(timeoutMillis: Long, maxItems: Int): Flow<List<T>> {
+    return channelFlow {
+        val buffer = mutableListOf<T>()
+        var bufferBeganAt = -1L
+
+        collectLatest { value ->
+            if (bufferBeganAt < 0) {
+                bufferBeganAt = System.currentTimeMillis()
+            }
+
+            buffer.add(value)
+
+            if (buffer.size < maxItems) {
+                // If the buffer is not full, wait until the time limit is reached
+                delay((System.currentTimeMillis() + timeoutMillis - bufferBeganAt).coerceAtLeast(0L))
```

Comment:
for some of the coroutine noobs (me) it might be nice to add a little comment, that this delay works with collectLatest, and delay is cancellable, so when `collectLatest` emits again, any currently running lambda/collector will cancel here.